ethdb/database.go

```diff
@@ -17,14 +17,24 @@
 // Package ethdb defines the interfaces for an Ethereum data store.
 package ethdb
 
-import "io"
+import (
+	"errors"
+	"io"
+)
+
+var (
+	// ErrKVNotFound is returned if a key is requested that is not found in
+	// the KeyValueStore.
+	ErrKVNotFound = errors.New("not found")
+)
 
 // KeyValueReader wraps the Has and Get method of a backing data store.
 type KeyValueReader interface {
 	// Has retrieves if a key is present in the key-value data store.
 	Has(key []byte) (bool, error)
 
 	// Get retrieves the given key if it's present in the key-value data store.
+	// Returns ErrKVNotFound if the key is not found.
```

Comment:
I am not sure this explicit error type is needed. 
Personally I feel it's OK to return the error if the element is not found in db, this error is mainly used for distinguishing `non-existent` element and `zero-value` element. For me it's too much to define the explicit error here.

Reply:
This is actually useful in at least two scenarios: caching and monitoring.

With a well defined datastore interface, it's possible to wrap said datastore with special purpose middleware to implement caching, sharding, logging, locking, metrics, combine different datastore with different characteristics ... which are all useful for heavy usage or running in production.

In the case of caching and monitoring, it's a requirement to distinguish a "not found" error (where the datastore is behaving properly) and an actual error. For example in the caching case, adding to / removing from the cache would happen instead of bubbling up an actual error.

That said, it's not necessary to have both `ErrKVNotFound` and `ErrSnapshotNotFound`, a single `ErrNotFound` would be enough. Would that be better?

Reply:
A single explicit error `ErrKVNotFound` is definitely better. I guess for all other errors like `ErrSnapshotNotFound`, `ErrSnapshotReleased` can be removed?

What's you guys opinion? cc @holiman @fjl

Reply:
Yes I'm OK with merging all not-found errors in ethdb.

Reply:
I merged those errors.